File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.*;

class Solution {
    public int solution(String str1, String str2) {

        Map<String, Integer> map1 = makeMap(str1);
        Map<String, Integer> map2 = makeMap(str2);

        // 둘 다 공집합이면 유사도 1
        if (map1.isEmpty() && map2.isEmpty()) {
            return 65536;
        }

        int intersection = 0;
        int union = 0;

        Set<String> keys = new HashSet<>();
        keys.addAll(map1.keySet());
        keys.addAll(map2.keySet());

        for (String key : keys) {
            int cnt1 = map1.getOrDefault(key, 0);
            int cnt2 = map2.getOrDefault(key, 0);

            intersection += Math.min(cnt1, cnt2);
            union += Math.max(cnt1, cnt2);
        }

        return (int) ((double) intersection / union * 65536);
    }

    private Map<String, Integer> makeMap(String str) {
        Map<String, Integer> map = new HashMap<>();
        str = str.toLowerCase();

        for (int i = 0; i < str.length() - 1; i++) {
            char a = str.charAt(i);
            char b = str.charAt(i + 1);

            if (Character.isLetter(a) && Character.isLetter(b)) {
                String key = "" + a + b;
                map.put(key, map.getOrDefault(key, 0) + 1);
            }
        }

        return map;
    }
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
function solution(str1, str2) {

    const makeMap = (str) => {
        const map = new Map();
        str = str.toLowerCase();

        for (let i = 0; i < str.length - 1; i++) {
            const a = str[i];
            const b = str[i + 1];

            if (a >= 'a' && a <= 'z' && b >= 'a' && b <= 'z') {
                const key = a + b;
                map.set(key, (map.get(key) || 0) + 1);
            }
        }
        return map;
    };

    const map1 = makeMap(str1);
    const map2 = makeMap(str2);

    // 둘 다 공집합이면 유사도 1
    if (map1.size === 0 && map2.size === 0) {
        return 65536;
    }

    let intersection = 0;
    let union = 0;

    const keys = new Set([...map1.keys(), ...map2.keys()]);

    for (const key of keys) {
        const cnt1 = map1.get(key) || 0;
        const cnt2 = map2.get(key) || 0;

        intersection += Math.min(cnt1, cnt2);
        union += Math.max(cnt1, cnt2);
    }

    return Math.floor((intersection / union) * 65536);
}
